### ISSUE TITLE
quick startup scripts not working, stuck in waiting room

### ISSUE THREAD

**Issue by davidschlangen** [clp-research/slurk#50]
Summary: The startup scripts do not seem to be working correctly for me. Server starts up, as do clients, but clients are then stuck in waiting room, stuck at "waiting for other player".

Steps to reproduce:
In terminal A: `python start_server_and_bot.py ../sample_bots/parrot.py `
In terminal B: `python start_clients.py`

Expected results: The client webrowsers open, clients are moved to a room where they can interact, and I can type something / demo an interaction.

Actual results: The clients do open, but they are forever stuck at "Waiting for other player..."

More info:
```
(py36) bash-3.2$ python start_server_and_bot.py ../sample_bots/turntaking/turntaking_bot.py 
Directory: /Users/das/mbp16_synced/work/svn/Gits/a_Projects/Projects/Infrastructure/slurk/slurk-myfork
Bots:  ['/Users/das/mbp16_synced/work/svn/Gits/a_Projects/Projects/Infrastructure/slurk/slurk-myfork/sample_bots/turntaking/turntaking_bot.py'] 

server started
starting /Users/das/mbp16_synced/work/svn/Gits/a_Projects/Projects/Infrastructure/slurk/slurk-myfork/sample_bots/turntaking/turntaking_bot.py 
token: a16bc3f7-fd30-4376-98e7-2b3f420bb42e
TTBot connected
loading layout from /Users/das/mbp16_synced/work/svn/Gits/a_Projects/Projects/Infrastructure/slurk/slurk-myfork/app/main/../static/layouts/waiting_room.json
TTBot joined room: Waiting Room
message handler error
Traceback (most recent call last):
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/engineio/server.py", line 405, in _trigger_event
    return self.handlers[event](*args)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/socketio/server.py", line 518, in _handle_eio_message
    self._handle_connect(sid, pkt.namespace)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/socketio/server.py", line 421, in _handle_connect
    self.environ[sid]) is False:
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/socketio/server.py", line 490, in _trigger_event
    return self.handlers[namespace][event](*args)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/flask_socketio/__init__.py", line 243, in _handler
    *args)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/flask_socketio/__init__.py", line 624, in _handle_event
    ret = handler()
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/flask_login/utils.py", line 260, in decorated_view
    return current_app.login_manager.unauthorized()
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/flask_login/login_manager.py", line 160, in unauthorized
    abort(401)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/werkzeug/exceptions.py", line 707, in abort
    return _aborter(status, *args, **kwargs)
  File "/Users/das/anaconda2/envs/py36/lib/python3.6/site-packages/werkzeug/exceptions.py", line 687, in __call__
    raise self.mapping[code](*args, **kwargs)
werkzeug.exceptions.Unauthorized: 401 Unauthorized: The server could not verify that you are authorized to access the URL requested.  You either supplied the wrong credentials (e.g. a bad password), or your browser doesn't understand how to supply the credentials required.
```

(That's actually new... Before it started without errors.)

**Comment by simeonschuez** [comment]
Regarding the error message: 

I just cloned dsg-bielefeld/slurk and set up a new virtual environment. I get the same error message, but _only_ if I want to connect the pairup bot. There are no errors for any other bot (including the TTBot). 

However, I don't think the error is due to the startup script, since I get the same error message if I try to connect the pairup bot manually (i.e. run the chat.py, generate a token using the web interface and run the bot in a separate terminal).

**Comment by davidschlangen** [comment]
@TimDiekmann , any ideas?

**Comment by TimDiekmann** [comment]
I'll look into this tomorrow. My first thought was, that the pairup bot gets a wrong token, but this doesn't seem to be the case.

**Comment by TimDiekmann** [comment]
I cannot really reproduce the error. Steps I made to reproduce:
- clone the repo
- configure _config.ini_
- `python start_server_and_bot.py ../sample_bots/parrot.py`
- `python start_clients.py`

The clients are also "stuck" in the waiting room, however when I call
```
python start_server_and_bot.py ../sample_bots/parrot.py ../sample_bots/pairup_bot.py
```
every client get paired.

My config:

```ini
[server]
secret-key = ysedcftgb
host = 127.0.0.1
port = 5001
database = botsi.db
logfile-date-format = %Y-%m-%d %H-%M-%S

[ssl]
enabled = no
certificate = PATH_TO_CERT_GOES_HERE
key = PATH_TO_KEY_GOES_HERE

[templates]
login-title = slurk - Login
chat-title = slurk - Chat
chat-header = slurk - Chatroom
token-title = slurk - Token generator

[client]
# Latency in ms
refresh-threshold = 800
# Wait at least X seconds until refresh
refresh-start = 5
# refresh after X seconds
refresh-max = 5
# number of ping-pong latency checks
ping-pong-latency-checks = 5
# show the latency by default
show-latency = yes
# show users by default
show-users = no

[tools]
browser1 = firefox
browser2 = firefox
client_names = "a,b,c,d"
```

Is your problem solved by just calling this?:
```
python start_server_and_bot.py ../sample_bots/parrot.py ../sample_bots/pairup_bot.py
```

If not, could you post your _config.ini_?

**Comment by TimDiekmann** [comment]
Regarding the unauthorizied problem: This is known and a feature. It occurs iff you start the server and there were some old connections to that server. Example:
- start server
- join server with bot or client
- stop server (don't quit bot or client)
- start server -> 401 Unauthorized for every client/bot which is still connected.

**Comment by davidschlangen** [comment]
This suggests then that killing the `start_server_and_bot` script doesn't properly kill the bot as well? Or is it the clients that are still active in the browser? I thought I had at least closed their browser windows -- does that still leave the clients active?

**Comment by simeonschuez** [comment]
> Regarding the unauthorizied problem: This is known and a feature. It occurs iff you start the server and there were some old connections to that server. 

I get that error right after rebooting the PC, so there shouldn't be any old connections.. I couldn't find anything suspicious using htop, too.
Also, I get this error message not until I try and connect the pairup bot (other bots seem to work fine) - shouldn't I get this error right after starting the server, if it was due to old connections?

**Comment by pkhdipraja** [comment]
I get the same error message regarding 401 Unauthorized when running the pairup bot following the instructions in the documentation (pairing up participants and running multiple rooms (and bots)). The error still persists even after rebooting and trying to run the instructions with and without virtual environment.

**Comment by TimDiekmann** [comment]
Could you list the exact steps you made?

**Comment by pkhdipraja** [comment]
Basically I just follow the exact instructions written in https://clp-research.github.io/slurk/slurk_multibots.html . Version of Python is 3.7.1

1) In Terminal A: `python chat.py`
2) Generate 2 tokens
3) In Terminal B: `python sample_bots/pairup_bot.py *token*` 

Here is also my config.ini : 
```
[server]
secret-key = A177308Z
host = 127.0.0.1
port = 5000
database = botsi.db
logfile-date-format = %Y-%m-%d %H-%M-%S

[ssl]
enabled = no
certificate = PATH_TO_CERT_GOES_HERE
key = PATH_TO_KEY_GOES_HERE

[templates]
login-title = slurk - Login
chat-title = slurk - Chat
chat-header = slurk - Chatroom
token-title = slurk - Token generator

[client]
refresh-threshold = 800
refresh-start = 5
refresh-max = 5
ping-pong-latency-checks = 5
show-latency = yes
show-users = no

[tools]
browser1 = firefox
browser2 = chromium-browser
client-names = John Zoidberg,Hubert Farnsworth
```

**Comment by TimDiekmann** [comment]
@simeonschuez I still cannot reproduce this. You mentioned, that you have found a way to solve this locally. Could you elaborate on this or create a PR?

**Comment by simeonschuez** [comment]
@TimDiekmann @pkhdipraja Yep, I think this problem is due to the login routine of the pairup bot (although I can't explain why I didn't have this problem a few months ago). I just noticed that the login mechanism of the pairup bot differs from all the other bots, and for me it's working if I change it in the same way. I can make a PR with the modified pairup bot - maybe today, otherwise I'll do it tomorrow in the morning.

**Comment by TimDiekmann** [comment]
Great, thanks!

**Comment by TimDiekmann** [comment]
Closing in favor of v2.0
